dissect/target/plugins/os/windows/wua_history.py

```diff
@@ -0,0 +1,879 @@
+    DATASTORE_UPDATE_TABLE = "tbHistory"
+
+    def __init__(self, target):
+        super().__init__(target)
+
+        self._datastore = None
+        datastore_path = target.fs.path(self.DATASTORE_PATH)
+        if datastore_path.exists():
+            try:
+                self._datastore = EseDB(datastore_path.open())
+            except Error as e:
+                self.target.log.warning("Error opening Windows DataStore database", exc_info=e)
+
+    def check_compatible(self) -> None:
+        if not self._datastore:
+            raise UnsupportedPluginError("No Windows DataStore found")
+
+    def get_table_records(self) -> Iterator[dict[str, RecordValue]]:
+        try:
+            table = self._datastore.table(self.DATASTORE_UPDATE_TABLE)
+        except KeyError:
+            return None
+
+        for record in table.records():
+            record_data = {}
+            for column in table.columns:
+                value = record.get(column.name)
+
+                if column.name == "Date":
+                    value = oatimestamp(value)
+                elif column.name in ("ServerId", "UpdateId"):
+                    value = decode_guid(value[:16]).encode("ascii")
+                elif column.name in ("MappedResult", "UnmappedResult"):
+                    value = hex(int(value) & 0xFFFFFFFF)
+                    if column.name == "MappedResult":
+                        if value in WUA_SUCCESS_CODES_MAP:
+                            record_data["result"] = "Succeeded"
+                            record_data["result_string"] = WUA_SUCCESS_CODES_MAP[value][0]
+                            record_data["result_description"] = WUA_SUCCESS_CODES_MAP[value][1]
+                        elif value in WUA_ERROR_CODES_MAP:
+                            record_data["result"] = "Failed"
+                            record_data["result_string"] = WUA_ERROR_CODES_MAP[value][0]
+                            record_data["result_description"] = WUA_ERROR_CODES_MAP[value][1]
+                        else:
+                            record_data["result"] = "Unknown"
+                            record_data["result_string"] = "Unknown error"
+                            record_data["result_description"] = "Unknown error"
+                elif column.name == "Title":
+                    kb = re.search(r"(KB.[0-9]*)", value)
+                    if kb:
+                        record_data["kb"] = kb.group(0)
+                record_data[column.name] = value
+            yield record_data
+
+    @export(record=WuaHistoryRecord)
+    def wua_history(self):
+        """Return the historical data of the Windows Update Agent from the DataStore.edb file.
+
+        Gives insight into the patch process and level of the system.
+        """
+        for record in self.get_table_records():
+            values = {TBHISTORY_NAME_MAP.get(key, key): value for key, value in record.items()}

```

Comment:
Is it intended to return the `key` when the `key` cannot be found in `TBHISTORY_NAME_MAP`?

Reply:
I grabbed this from the UAL plugin. I assumed it was written this way for a reason so I left it. Can change if desired.